src/main/scala/org/broadinstitute/dsde/workbench/sam/util/DistributedLock.scala

```diff
@@ -0,0 +1,105 @@
+package org.broadinstitute.dsde.workbench.sam.util
+import java.util.concurrent.{Executor, TimeUnit}
+
+import cats.effect.{IO, Resource, Timer}
+import cats.implicits._
+
+import scala.util.Try
+import com.google.cloud.firestore.{DocumentReference, DocumentSnapshot, Transaction}
+import org.broadinstitute.dsde.workbench.google.{CollectionName, Document, GoogleFirestoreOps}
+import org.broadinstitute.dsde.workbench.model.WorkbenchException
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.{FiniteDuration, _}
+
+object DistributedLock {
+  private[util] val EXPIRESIN = "expiresIn"
+  private[util] val ISLOCKED = "isLocked"
+  private[util] val EXPIRESINTIMEUNIT = TimeUnit.MILLISECONDS
+  implicit val cs = IO.contextShift(ExecutionContext.global)
+  implicit val timer = IO.timer(ExecutionContext.global)
+
+  val executor = new Executor {
+    override def execute(command: Runnable): Unit = ExecutionContext.global.execute(command)
+  }
+
+  def withLock(lockPath: LockPath)(implicit googleFirestoreOps: GoogleFirestoreOps[IO]): Resource[IO, Unit] = Resource.make{
+    retry(getLock(lockPath), 5 seconds, 5)
+  }(_ => releaseLock(lockPath))
+
+  private[util] def getLock[A](lockPath: LockPath)(implicit googleFirestoreOps: GoogleFirestoreOps[IO]): IO[Unit] =
+    googleFirestoreOps.transaction { (db, tx) =>
+      val docRef = db.collection(lockPath.collectionName.asString).document(lockPath.document.asString)
+      val res = for {
+        lockStatus <- getLockStatus(tx, docRef)
+        _ <- lockStatus match {
+          case Available => setLock(tx, docRef, lockPath.expiresIn)
+          case Locked => IO.raiseError(new WorkbenchException("can't get lock"))
+        }
+      } yield ()
+      res.unsafeRunSync()
```

Comment:
calling `unsafeRunSync` seems odd here. Why doesn't `googleFirestoreOps.transaction` that a function that produces `F[A]` instead of a naked `A`?

Reply:
```ApiFutures.addCallback(
            db.runTransaction(new Transaction.Function[A]{
              override def updateCallback(transaction: Transaction): A = ops(db, transaction)
            }),
            callBack(cb),
            executor
          )
```
This is the java interface, we have to get an `A` here because of `new Transaction.Function[A]` here...`new Transaction.Function[IO[A]]`  doesn't work (java lib probably doesn't understand it).

I think calling `.unsafeRunSync()` here should be okay since it's inside a callback we pass to the java lib, which gets executed asynchronously on java lib's execution context.